### PR TITLE
Have CRNG avail early on boot and maximize ligcrypt entropy sources/efficiency

### LINKED ISSUE #1476: CRNG not ready and entropy used early at boot
EDIT for current state:
- This PR enables CPU HW crypto backend for both AMD and Intel and passes `random.trust_cpu=1` from coreboot command line to Heads kernel boot option.
- Consequently, crng is ready prior of tpm calls and shredding operations which currently read from /dev/urandom which is not locking if not enough entropy available
- qemu depends on host's /dev/urandom through cirtio, and cannot depend on AMD cpu extension since emulated cpu is too old to offer RDRAND compatible extention.
- Next step could be to reenable TPM and JITTER as expected initially, and have rng-tools and licap-ng deployed under Heads.
- As of now the current approach of this PR is better than what was in master.
- Will try to find ways to compare both with available tools
- Still needed: see how libcrypt deals with entropy before generating key material (gnupg)

------
Older

Heads uses entropy early at boot through while CRNG initialization is not yet complete. 
This is the result of explorations done, goal being to open a debate on the proper way forward.

The problem is that the CRNG is not ready when some system services need to use random numbers at boot time, such as shredding files, calling tpm binary, generating secrets, disk unlock key, and most importantly generating in-memory key material. 

These operations are crucial for the security and privacy and they should not be compromised by a lack of entropy or a weak random number generator. The reason for this problem is that the kernel needs to collect at least 4096 bits of entropy before it can initialize the CRNG, and the default entropy sources are not sufficient or fast enough to provide that amount of entropy at boot time.

To solve this problem, I have explored some possible options to increase the entropy sources and reduce the crng readiness time. These options are:

- Use the kernel option `random.trust_cpu=1` and pass it as a command line parameter to skip the PRNG part of the boot process and use the CPU's hardware random number generator (e.g. RDRAND for x86 CPUs) as a source of randomness, without mixing it with other entropy sources. This will reduce crng readiness time by avoiding the entropy starvation problem. However, this may not be secure or trustworthy, as there is a possibility of a backdoor in the RDRAND implementation or the XOR instruction that could compromise the quality of randomness. Some experts recommend using RDRAND only as an additional entropy source, not as the sole one. 
- Use the TPM chip as another source of entropy, as it can provide high-quality randomness that is resistant to physical attacks and tampering.  You can check if the TPM is recognized as a hardware random number generator (HRNG) by running `cat /proc/sys/kernel/random/entropy_avail`.
- Use other sources of entropy like jitter, hardware interrupts, mouse movements, keyboard presses. Linux uses a complex algorithm to combine multiple entropy sources into a single pool, and then hashes the pool to produce random output. Mixing entropy sources can improve the security and robustness of randomness, especially if some of the sources are compromised or biased.
- If we decide to trust CPU extention for crng early readiness, we could deactivate it early in Heads init script by changing the kernel parameter `random.trust_cpu` after the crng is ready.  This way, we could use the CPU as the sole source of entropy for CRNG initialization, but switch to other sources afterwards. 

To enable these options, we need to make some changes to the kernel configuration and command line parameters. These changes are:

- To enable RDRAND as an entropy source, we need to set `CONFIG_ARCH_RANDOM=y` in the kernel configuration, and add `random.trust_cpu=1` to the kernel command line.
- To enable TPM as an entropy source, we need to set `CONFIG_HW_RANDOM_TPM=y` in the kernel configuration, and add `rng_core.default_quality=1000` to the kernel command line.
- To enable jitterentropy as an entropy source, we need to set `CONFIG_CRYPTO_JITTERENTROPY=y` in the kernel configuration, and add `kernel.jitterentropy.ll=10`, `kernel.jitterentropy.ml=10`, `kernel.jitterentropy.bs=64`, and `kernel.jitterentropy.bc=16` to the kernel command line. These parameters control how much jitter is generated by different methods.
- To enable external devices as entropy sources, you need to set `CONFIG_HW_RANDOM=y` in the kernel configuration.

I would like to hear your opinions and feedback on these options, and discuss which one is the best way forward for Heads.
As of today, crng is ready after tpm binary is called and after shredding some files. If one user is quick selecting menu options, the crng might not be ready when setting a TPM disk unlock key, while really improbable.

The goal here is to try to have crng ready as early as possible. Another option would be to wait for the entropy pool to full prior of leaving Heads init, but blocking is not pretty nor practical.

I propose:
- To not trust cpu extension first (requires coreboot config to pass explicitely `random.trust_cpu=1` but impement all other changes and see how that goes.
- If crng is still not ready on real hardware when tpm and shredding calls happen, revisit the precedent option and modify init script to remove trust on cpu when "inside" of Heads (so aeraly tpm unseal operations and shredding of files happen on a crng that is ready.

@JonathonHall-Purism @saper others? comments?

---

Some sources:
- Intel's RDRAND and RDSEED: A Cautionary Tale: https://www.cryptologie.net/article/487/intels-rdrand-and-rdseed-a-cautionary-tale/
- Intel's RDRAND: A Backdoor?: https://www.schneier.com/blog/archives/2013/10/intels_rdrand.html
- RDRAND: The Case for Trusting CPU Hardware Random Number Generators: https://www.intel.com/content/www/us/en/developer/articles/technical/rdrand-the-case-for-trusting-cpu-hardware-random-number-generators.html

### PR DESCRIPTION
**Pull request by tlaurion**
This PR at time of merging:

- unifies linux configs and make sure HW entropy is on
- unified coreboot kernel options to set log default to loglevel=2
- Output of warn/die/DEBUG/TRACE are now under kernel log if DEBUG is enabled
- crng is ready early giving weight to HW provided entropy source, removing errors on dmesg and console

---

Older traces of history:
- https://github.com/osresearch/heads/pull/1478#issuecomment-1747016492 clarifies what trust cpu is really about. In short, it is about giving weight to RDRAND RDSEED and not giving warning about crng and accessing random pools prior of them having enough entropy collected.
-  https://github.com/osresearch/heads/pull/1478#issuecomment-1747018547 shows that HOTP/TOTP boards have differences in crng init being ready at different times (4s-30 seconds) depending of if USB stacks related modules loaded and IO linked to all of those operations. We could duplicate linux configs to trustcpu on TOTP boards only, but that doesn't make sense following https://github.com/osresearch/heads/pull/1478#issuecomment-1747016492 analysis.
- It is thought that enabling TRUSTCPU is better then nothing, while it is to be reminded that even if trustcpu is not enabled, CPU RDRAND/RDSEED are used where available to fill the pool. It is also noted that we would have rng-tools to mix entropy under /dev/urandom.
- As of now, /dev/urandom is used directly, even if pool entropy is available or not by shred, tpm cryptsetup and others. But, those calls are made when the pool is having enough entropy anyway.
- libgcrypt by this PR makes all available optimizations for gnupg calls. If jitter was enabled in the kernel, libgrcypt would use it. 

So basically, next steps for future work are:
- Add rng-tools to mix entropy source and make that avail under /dev/urandom which is used by most tools
- Enable jitter in all kernel configuration so that IO/cpu + RDRAND/RDSEED+Jitter is used at time of generating keys

But as of now, simply trusting CPU is good enough and if not, shred and tpm calls made prior of crng being ready are also good, while not easthetic.
- 

Note: https://github.com/osresearch/heads/commit/00e715660a7a85662e6df3b272ac939461082cce#diff-a890ddf36ae7dca110b70df101c97525a4d8265212ebed3f92562d47162af3a2 enabled trustcpu on most older boards.

- [ ] Needed for sandy/ivy otherwise crng is not ready prior of pool being used (takes 16s-30 seconds prior of dmesg showing `random: crng init done`
- [ ] Needed for t440p?
- [ ] Needed for other platforms?

Consequence as this PR goes through is that
- cryptsetup reads through /dev/urandom which is not blocking if pool is not ready
- shred and tpm are using /dev/urandom as well which is unblocking as opposed to /dev/random
- gnupg currently uses kernel entropy and sets base for additional entropy sources and accelerations per removal --disable-asm under libgcrypt

Older:
-----
This is PoC to fix #1476 

- ~**Does rely on trust CPU**. CPU RAND is first source to fill entropy pool early on boot. coreboot's linux command line turns it "on" explicitly~ edit: not needed on qemu with virtio supported in kernel and qemu/kvm existing command line options reusing host's available entropy as source.
- ~Jitter is added into kernel configs, and coreboot's linux command line option sets basic values to configure jitter~ Tpm is now turned off since not aggregated by kernel alone
- ~TPM was already configured to be a source of randomness in kernel config, but weight was not adjusted in coreboot's kernel config to ponder properly. This is done now~ Jitter and TPM hardware source of entropy are now turned off since not aggregated by kernel alone.
- libgcrypt is now compiled to take advantage of available CPU enhancements, including jitter and RDRAND if available, where jitter is used by libgrypt in gnupg operations. This is to be tested, but master was disabling all of those by configuring libgcrypt with --disable-asm
- qemu now turn on virtio "HW" entropy source (virtio, so host's provided /dev/urandom). ~Comparable to RDRAND
- Not clear what is available out of the box for PPC

Tested in QEMU (TCG, not KVM, sloooooooow)
- ~crng init=0 to readiness takes less then a second.~ with virtio, crng on qemu is ready under 6s, prior of first entropy pool access.
- DIFFERENT for ivy bridge+ (xx30: x230/t430/t530/w530/t440p and comparable). **crng is ready between 16s-30s**.
- ~**Sandy bridge- (xx20: x20/t420/others) will continue to be slow to have crng ready at boot: RDRAND was added to ivy, not sandy: so xx20 Lenovo will rely on jitter from libgcrypt here even more. Thanks to libgcrypt 1.10+ this is on by default.**~

TODO:
- [x]  Validate that the sources of entropy are available, selected and used
- [x] ~Adapt linux config to make sure CPU features are exposed (AMD/Qemu)~ Edit: virtio only here, amd cpu is older family and needed RDSEED nor RDRAND is available on q35.
- [ ] Give traces and testing paths for reproducibility and auditability of change. Edit: gpg --benchmark not available. Edit: even if availabe, not su useful.
- [x] Validate that libgcrypt uses additional entropy source when generating keys in memory: edit: this is sufficient AFAIK without need to trust CPU to have crng and with libgcrypt enabling and mixing all available platform enhancements.  
- [ ] check what is available to PPC

### CODE REVIEW

**Comment by tlaurion** [comment]
No. Won't fly without rngd. Kernel doesn't aggregate multiple sources. Hopefully, rngd that depends on libgrypt we already use for gpg2 can be used and found resources stipulate that it could be as small as 18k when compiled with -Os. Let's see.

**Comment by tlaurion** [comment]
rngd-tools have dependencies on libcap which hardcodes cross compilation assumptions under Makefile and Make.Rules... Not successful as of now.

Can we remove libcap dependencies? Seems like another subproject.


As in current state
- qemu depends on virtio which is fed by host. Good enough.
- other physical platforms could depend on either jitter, cpu trust or TPM, where tpm seems to require additional tooling.

We could probably trust libgcrypt + cpu extension here as a start.

**Comment by tlaurion** [comment]
Ok so after having looked into this deeper:
- Heads taps into /dev/urandom so not locking if not enough entropy. Requires pool to be ready prior of generating key material and such
- Kernel does not aggregate multiple sources of entropy, offloading the task to userland (rng-tool)
- My past attempt to integrate libcap was not libcap-ng but I think we should take gradual steps here
- tpm hw needs toolstack to read into it, but source is slow (10-20 bits second)
- jitter needs toolstack to read into it, could be alternative
- cpu hw (AMD/INTEL) is fast enough. 

I added sysctl fs in kernel for all boards. Enabled CPU HW for all boards. Qemu uses virtio of host.
So as of now, crng is ready prior of doing anything early )shred of secrets, tpm operations) which is better then before.

I have not digged into libgcrypt entropy sourcing, but I also think that it will be better then before. Would need testing here, but I think the PR is an improvement prior of before with next commits.

Will clean after review.


@JonathonHall-Purism Haven't touched librems outside of putting librem_l1um2 to oldconfig (linux.save_in_old_config would require a lot of review, you are activating a bunch of unneeded stuff right now). Let me know if you want me to enable comparative configs for librems in this PR.

This is needed prior of in-memory key gen PR to be testable.

**Comment by tlaurion** [comment]
By disabling `--disable-asm` in libgcrypt 1.10.1:
```
config.status: executing gcrypt-conf commands
         
        Libgcrypt v1.10.1 has been configured as follows:
         
        Platform:                  GNU/Linux (x86_64-pc-linux-musl)
        Hardware detection module: libgcrypt_la-hwf-x86
        Enabled cipher algorithms: arcfour blowfish cast5 des aes twofish
                                   serpent rfc2268 seed camellia idea salsa20
                                   gost28147 chacha20 sm4
        Enabled digest algorithms: crc gostr3411-94 md4 md5 rmd160 sha1
                                   sha256 sha512 sha3 tiger whirlpool stribog
                                   blake2 sm3
        Enabled kdf algorithms:    s2k pkdf2 scrypt
        Enabled pubkey algorithms: dsa elgamal rsa ecc
        Random number generator:   default
        Try using jitter entropy:  yes
        Using linux capabilities:  no
        FIPS module version:       
        Try using Padlock crypto:  yes
        Try using AES-NI crypto:   yes
        Try using Intel SHAEXT:    yes
        Try using Intel PCLMUL:    yes
        Try using Intel SSE4.1:    yes
        Try using DRNG (RDRAND):   yes
        Try using Intel AVX:       yes
        Try using Intel AVX2:      yes
        Try using ARM NEON:        n/a
        Try using ARMv8 crypto:    n/a
        Try using PPC crypto:      n/a
```

To support PPC crypto, it seems we will need libcap as well later on. :/ another point for rng-tools (which also depends on libcap-ng)

**Comment by tlaurion** [comment]
Todo
- [ ] check output of configure for Talos and validate detection are enabled

**Comment by tlaurion** [comment]
@saper @JonathonHall-Purism @daringer On second thoughts, i'm not even convinced that `--disable-asm` should be removed from libgcrypt.

I mean as shown at https://github.com/osresearch/heads/pull/1478#issuecomment-1699424109, with or without asm code to detect CPU enabled acceleration for different CPUs extensions (which speedups in memory keygen) I want to bring your attention there on he fact that what we really want here, which is `Try using jitter entropy:  yes` is already enabled by default on libgcrypt 1.10.1 as we currently use it.

So my reasoning here is that by trusting the CPU for crng init (early boot entropy pool ready), calling gnupg already takes advantage of jitter to not only use flimsy entropy as in master, or now trusted  HW provided RDRAND cpu extension enforced entropy.

Note that we have the space to enable all of that. Question is if `--disable-asm` was  put in place early under Heads libgcrypt for reproducibility reason (I would doubt so but maybe) which is not a problem for a while now.

Thoughts? I think that in current state, this PR is a massive inprovements where next steps would be to attack rng-tools inclusion (meaning libcap-ng inclusion) if we want to enable jitter in kernel + tpm hw entropy backend and have rng-tool feed /dev/random and /dev/urandom with proper entropy.

But in my opinion, this is not needed here and now and can be enabled later on in future PR.
Putting PR as ready for review.

**Comment by tlaurion** [comment]
Ok, after some challenging discussions with @JonathonHall-Purism :

New facts:
- cryptsetup uses /dev/urandom by default, meaning it trusts totally entropy provided by RDRAND without mixing it in with anything else. To change behavior, one would have to call cryptsetup with --use-random to generate long-term volume key, use at both luksFormat and by cryptsetup-reencrypt.
- libgcrypt 1.10+ changed their defaults to mix in jitter in kernel provided source, so in memory keygen is ok currently on master, no need to trust cpu.
- Need is to review current crng status on x230, observations from qemu are not valid
- qemu will keep trust cpu, since virtio "hw" source is actually tapping into the host's entropy pool which delegates to the host managing proper entropy pool (where deamons are normally present)
-  Heads taps into /dev/urandom for shred and tpm calls early on boot. It is ok if crng is ready, would be better is not /dev/urandom but that would block.
- Kernel has its own basic entropy gen under crng. Read comments for crng and entropy collection routines at https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/char/random.c

Get homes messages:
- We are better without HW RDRAND if crng is ready on boot
- gpg in memory keygen is already injecting entropy whike using libgcrypt 1.10+ by default
- removing --disable-asm is permitting libgcrypt to use what is available to mix what is available on the host without trusting completely the CPU HW rand


So next commits will
- disable trusting CPU completely from coreboot command line to Heads payload, making the hw source availabile but not enforcing it
- qemu board will keep trust cpu since it trusts virtio, hence the host's source of entropy as being "hw"

**Comment by JonathonHall-Purism** [comment]
Thanks for summing that up @tlaurion.  For context, I did some quick checks on my 15v4 and it appeared that crng was already initialized before it was needed on that device from kernel sources, so there is no issue there.  I suspect this is only an issue on qemu, but we need to confirm on other devices as well.

We observed as you noted that cryptsetup uses the kernel random number generator; we need this to be a good source of entropy.  While our sources of entropy at this point are limited, I would prefer not to trust RDRAND if we do not have to.  It appears to me from initial tests that we are getting enough entropy from existing sources outside of qemu, so if there are some devices where we aren't getting that entropy soon enough, maybe there is another solution.

For qemu I am totally fine with adding the necessary bits to get initial entropy from the host, including trust_cpu if needed.

>cryptsetup [...] trusts totally entropy provided by RDRAND without mixing it in with anything else. To change behavior, one would have to call cryptsetup with --use-random [...]

Nitpick, --use-random does not affect cryptsetup's trust of kernel RNG, it trusts it either way.  The difference is only that /dev/random will block before crng is ready, while /dev/urandom just returns low-entropy numbers immediately.  It looks like cryptsetup in Heads is configured to use urandom by default (from cryptsetup --help).

**Comment by tlaurion** [comment]
@JonathonHall-Purism forced push to only change qemu linux config. crng is ready within 5s and before entropy is expected. No more error on qemu side. 

Please review, this will be smaller steps and will extensively test this as base for forthcoming PRs.

**Comment by tlaurion** [comment]
Bad news here. TRUST CPU was enabled for a while which is why crng was not complaining on crng for a lot of platforms on which that change was reverted under https://github.com/osresearch/heads/pull/1478/commits/f59a63a96b1907d6153b23931a8b19979332d536

@JonathonHall-Purism can you verify that builds for that commit (without trust on CPU) has a crng ready before tpm and shred calls?

on x230 without TRUST CPU:
signal-2023-10-03-140301


We can see TPM and shred calls happening on empty pool prior of crng being ready.


So I guess I will revert this for xx20/xx30 so that TRUST CPU is on and state it clearly on board configs?

**Comment by tlaurion** [comment]
Letting the build finish prior of pushing more commit.

The following needs to be added to fix mounting devpts over /dev/pts:
```
diff --git a/config/linux-c216.config b/config/linux-c216.config
index 7bccfba021a..662125a49a5 100644
--- a/config/linux-c216.config
+++ b/config/linux-c216.config
@@ -1375,7 +1375,7 @@ CONFIG_CONSOLE_TRANSLATIONS=y
 CONFIG_VT_CONSOLE=y
 CONFIG_HW_CONSOLE=y
 CONFIG_VT_HW_CONSOLE_BINDING=y
-# CONFIG_UNIX98_PTYS is not set
+CONFIG_UNIX98_PTYS=y
 # CONFIG_LEGACY_PTYS is not set
 # CONFIG_LDISC_AUTOLOAD is not set
 
diff --git a/config/linux-librem_common-6.1.8.config b/config/linux-librem_common-6.1.8.config
index 59ffef787b4..b7e7c76130e 100644
--- a/config/linux-librem_common-6.1.8.config
+++ b/config/linux-librem_common-6.1.8.config
@@ -141,7 +141,6 @@ CONFIG_E1000E=m
 # CONFIG_WLAN is not set
 # CONFIG_INPUT_MOUSE is not set
 # CONFIG_SERIO_SERPORT is not set
-# CONFIG_UNIX98_PTYS is not set
 # CONFIG_LEGACY_PTYS is not set
 CONFIG_SERIAL_8250=y
 # CONFIG_SERIAL_8250_DEPRECATED_OPTIONS is not set
diff --git a/config/linux-librem_common.config b/config/linux-librem_common.config
index c7a0d2175b5..d718ce5370a 100644
--- a/config/linux-librem_common.config
+++ b/config/linux-librem_common.config
@@ -1385,7 +1385,7 @@ CONFIG_CONSOLE_TRANSLATIONS=y
 CONFIG_VT_CONSOLE=y
 CONFIG_HW_CONSOLE=y
 CONFIG_VT_HW_CONSOLE_BINDING=y
-# CONFIG_UNIX98_PTYS is not set
+CONFIG_UNIX98_PTYS=y
 # CONFIG_LEGACY_PTYS is not set
 CONFIG_LDISC_AUTOLOAD=y
 
diff --git a/config/linux-nitropad-x.config b/config/linux-nitropad-x.config
index 7f3fbbf3617..c82992c8dc0 100644
--- a/config/linux-nitropad-x.config
+++ b/config/linux-nitropad-x.config
@@ -1483,7 +1483,7 @@ CONFIG_CONSOLE_TRANSLATIONS=y
 CONFIG_VT_CONSOLE=y
 CONFIG_HW_CONSOLE=y
 CONFIG_VT_HW_CONSOLE_BINDING=y
-# CONFIG_UNIX98_PTYS is not set
+CONFIG_UNIX98_PTYS=y
 # CONFIG_LEGACY_PTYS is not set
 CONFIG_LDISC_AUTOLOAD=y
 
diff --git a/config/linux-t440p.config b/config/linux-t440p.config
index c8b0fbe9398..74ff2fe57b1 100644
--- a/config/linux-t440p.config
+++ b/config/linux-t440p.config
@@ -1405,7 +1405,7 @@ CONFIG_CONSOLE_TRANSLATIONS=y
 CONFIG_VT_CONSOLE=y
 CONFIG_HW_CONSOLE=y
 CONFIG_VT_HW_CONSOLE_BINDING=y
-# CONFIG_UNIX98_PTYS is not set
+CONFIG_UNIX98_PTYS=y
 # CONFIG_LEGACY_PTYS is not set
 CONFIG_LDISC_AUTOLOAD=y
 
diff --git a/config/linux-x230-flash.config b/config/linux-x230-flash.config
index dc64463ff13..f3d1fb0e863 100644
--- a/config/linux-x230-flash.config
+++ b/config/linux-x230-flash.config
@@ -1071,7 +1071,7 @@ CONFIG_CONSOLE_TRANSLATIONS=y
 CONFIG_VT_CONSOLE=y
 CONFIG_HW_CONSOLE=y
 CONFIG_VT_HW_CONSOLE_BINDING=y
-# CONFIG_UNIX98_PTYS is not set
+CONFIG_UNIX98_PTYS=y
 # CONFIG_LEGACY_PTYS is not set
 # CONFIG_LDISC_AUTOLOAD is not set
 
diff --git a/config/linux-x230-legacy.config b/config/linux-x230-legacy.config
index f90e983d699..54b9cdc2886 100644
--- a/config/linux-x230-legacy.config
+++ b/config/linux-x230-legacy.config
@@ -1181,7 +1181,7 @@ CONFIG_CONSOLE_TRANSLATIONS=y
 CONFIG_VT_CONSOLE=y
 CONFIG_HW_CONSOLE=y
 CONFIG_VT_HW_CONSOLE_BINDING=y
-# CONFIG_UNIX98_PTYS is not set
+CONFIG_UNIX98_PTYS=y
 # CONFIG_LEGACY_PTYS is not set
 # CONFIG_LDISC_AUTOLOAD is not set
 
diff --git a/config/linux-x230-maximized.config b/config/linux-x230-maximized.config
index cd46aad1e24..b932b05b5b8 100644
--- a/config/linux-x230-maximized.config
+++ b/config/linux-x230-maximized.config
@@ -1372,7 +1372,7 @@ CONFIG_CONSOLE_TRANSLATIONS=y
 CONFIG_VT_CONSOLE=y
 CONFIG_HW_CONSOLE=y
 CONFIG_VT_HW_CONSOLE_BINDING=y
-# CONFIG_UNIX98_PTYS is not set
+CONFIG_UNIX98_PTYS=y
 # CONFIG_LEGACY_PTYS is not set
 # CONFIG_LDISC_AUTOLOAD is not set
```

**Comment by tlaurion** [comment]
Freezing this PR for the moment, continuing tests under https://github.com/tlaurion/heads/tree/stenghten_entropy_sources_with_jitter_and_TPM-withTrustCPU which trustcpu is enabled as in master because of https://github.com/osresearch/heads/commit/00e715660a7a85662e6df3b272ac939461082cce#diff-a890ddf36ae7dca110b70df101c97525a4d8265212ebed3f92562d47162af3a2R1472 but when disabled, sandy/ivy and probably other older architectures read entropy through an empty crng pool....

**Comment by JonathonHall-Purism** [comment]
@tlaurion and I both researched this and discussed.

In practice, TRUST_CPU essentially only controls whether the warning about the pool not being fully initialized is printed.  Linux uses RDRAND/RDSEED either way, the only difference is whether it credits any entropy.  With TRUST_CPU=n it uses them but credits no entropy, the assumption being that a bad entropy source could not be worse than not using it at all.

Currently in Heads I feel that this warning has little value.  It's currently caused by shred most of the time, and since we are operating on tmpfs, overwriting the data with any value is sufficient.  Key generation occurs after plenty of I/O and user input and there appears to be sufficient entropy at that point regardless of TRUST_CPU.

There is room for debate and improvement, but I think the status quo is improved just by enabling TRUST_CPU to remove a warning that isn't useful in this context.

Some thoughts for future work (will open issues):
* Re-examine our usage of shred - perhaps overwrite data a different way that doesn't use CRNG
* Add additional sources of entropy (CPU jitter, etc.), which requires an RNG daemon
* Question the assumption that a bad RDRAND/RDSEED isn't worse than not using it at all (if it is malicious instead of just bad)

**Comment by tlaurion** [comment]
Ok. Further testing shows a big difference in CNRG readiness on time, depending of if the board is HOTP/TOTP only.
The reasoning is that early on boot, if HOTP is used, USB stack kernel modules are loaded early, same for USB keyboard, and then unseal-hotp is called and some related additional IO are being made (USB stack modules measured and loaded) which are taken into consideration for entropy pool.

Testings shows that even on TOTP boards (non-HOTP), doing factory-reset or manual gpg operations on command line will have crng ready prior of important operations.

It is also to be noted that errors given by shred call is benign, since those calls are wiping files in ramfs which blocks would be rewritten with not so random bytes, still overwiting secret intended to be wiped.

The only warning i'm still worry-ish about is tpm calls when crng is not ready:
signal-2023-10-04-095543

**Comment by tlaurion** [comment]
@JonathonHall-Purism that was harder then I thought.

Some notes: 
- config/linux-librem_common-6.1.8.config is still in defconfig, which hides that "CONFIG_RANDOM_TRUST_CPU=y" is set. You can check for yourself with `make BOARD=librem_14 linux.modify_and_save_oldconfig_in_place`. Note that a lot of modules are compiled even if unused, that CPU support is generic and does not turn on CPU optimizations as per other boards that were tuned in oldconfig format. I would suggest to compare and put in oldconfig asap. I also added CONFIG_UNIX98_PTYS=y to get rid of warning on pts early on boot (was also missing under qemu linux config).

As of now, this PR is a bit messy, adds cttyhack to call BOOTSCRIPT otherwise, dual console setups (qemu) was reporting to not have proper tty when tty is called. Thoughs on that welcome.


Last changes add preoper DEBUG/TRACING output under dmesg and in console. Reason why those were intertwined by default under qemu is because coreboot config's kernel command line was passing "debug". To have this enabled to have Heads DEBUG/TRACE/die/warn messages intertwined under console output, dmesg -n 8 is needed. But detection of if CONSIF_DEBUT_OUTPUT is setuped either on board config or through config.user on-demand config requires changes under init and proper different output under init.

All of this to permit to detect tpm call on qemu setup, which is the changes coming next if VIRTIO HW source is not enabled in kernel and have all rand calls from unready pool.

To accomplish the same on real hw, the board config needs to enable DEBUG_CONSOLE_OUTPUT in board config otherwise early debug messages are not intertwined since the checks for DEBUG_CONSOLE_OUTPUT are not positive until combine_configs happens.


All unneeded commits will be squashed, of course....

**Comment by tlaurion** [comment]
This is output of dmesg from https://github.com/osresearch/heads/pull/1478/commits/dabdc5847e41849f2c11796e10e10abbc9dcd197

[dmesg_no_hw_trust.txt](https://github.com/osresearch/heads/files/12833435/dmesg_no_hw_trust.txt)

**Comment by tlaurion** [comment]
@JonathonHall-Purism This is the output of https://github.com/osresearch/heads/pull/1478/commits/7a084741d3d6bfaf9db255860fdb170434a7b408 for HOTP without HW entropy nor CPU trust

[dmesg_hotp_no_hw_trust.txt](https://github.com/osresearch/heads/files/12833618/dmesg_hotp_no_hw_trust.txt)

**Comment by tlaurion** [comment]
This is the output of https://github.com/osresearch/heads/pull/1478/commits/f50c29615e4cf24f721cf66323abb701a3a368c3 with hw entropy, no CPU trust resealing TOTP/HOTP
[dmesg_hotp-with_hw-no_trust.txt](https://github.com/osresearch/heads/files/12833749/dmesg_hotp-with_hw-no_trust.txt)

**Comment by tlaurion** [comment]
@JonathonHall-Purism with https://github.com/osresearch/heads/pull/1478/commits/adb5808ecce555251f0858084f71996baa49d3dc we add CPU trust to intel HW entropy source which gives this at totp+hotp sealing 
[dmesg_hotp-with_hw-with_trust.txt](https://github.com/osresearch/heads/files/12833822/dmesg_hotp-with_hw-with_trust.txt)

**Comment by tlaurion** [comment]
@JonathonHall-Purism https://github.com/osresearch/heads/pull/1478/commits/9cdf0856e196f88091d968283a96f1ea86bab49a adds libgcrypt benchmark app which is ~50k uncompressed.

**Comment by tlaurion** [comment]
Basically, I think I will stage this pr and
- keep linux config changes here
- init changes
- ash_functions change

Opinion on cttyhack needed, can also be separated from here.

And move the rest to other PRs.
@JonathonHall-Purism good with you? 
- gnupg toolstack version bump was required to correct bugs under yubikey
- board config changes and coreboot files changes are requires to remove igfx related stuff that don't matter anymore (where dgpu boards still repend on it and most are UNTESTED now)

**Comment by tlaurion** [comment]
Squashing

File: modules/libgcrypt [hunk 1]
```diff
@@ -14,7 +14,6 @@ libgcrypt_configure := \
 	--prefix "/" \
 	--disable-static \
 	--with-gpg-error-prefix="$(INSTALL)" \
-	--disable-asm \
```
Comment (JonathonHall-Purism) [inline modules/libgcrypt:17]: Summarizing a discussion via DM - I'm good with this change.

And odd thing sticking out to me was that libgcrypt has code to use RDRAND directly for up to 25% of its entropy, which would have been prevented in the past with `--disable-asm`.  That doesn't make much sense with the kernel already including RDSEED in its CRNG, and it doesn't even know about things like the broken Ryzen microcode, etc.  But this code is no longer active by default, that code is for USE_RNDOLDLINUX, default is now USE_RNDGETENTROPY.

So no issues here, let's keep this change.

File: initrd/etc/ash_functions [hunk 2]
```diff
@@ -4,26 +4,34 @@
...
 
 warn() {
-	echo >&2 " *** WARNING: $* ***";
+	if [ "$CONFIG_DEBUG_OUTPUT" = "y" ];then
+		echo " *** WARNING: $* ***" | tee -a /tmp/debug.log /dev/kmsg > /dev/null;
+	else
+		echo >&2 " *** WARNING: $* ***";
+	fi
 	sleep 1;
 }
```
[the remark this answers is not in the document]
Comment (JonathonHall-Purism) [reply initrd/etc/ash_functions:23]: Nevermind, @tlaurion clarified to me that this is due to the interaction with enabling kernel messages to console (`dmesg -n 8`).  If we do both, then these messages show up on the console _twice_, which is pointless.

**Comment by JonathonHall-Purism** [comment]
@tlaurion Thanks for clarifying both of those last discussions with me, let's merge :+1: :ship:

**Pull request merged**